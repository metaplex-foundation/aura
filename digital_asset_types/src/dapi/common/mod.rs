mod asset;
pub use asset::*;
