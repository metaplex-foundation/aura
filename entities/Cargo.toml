[package]
name = "entities"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
serde = "1.0.193"
solana-sdk = "~1.16"
num-derive = "0.4.1"
num-traits = "0.2.17"
schemars = "0.8.6"
serde_derive = "1.0.195"
sqlx = { version = "0.6.2", features = ["macros", "runtime-tokio-rustls", "postgres", "uuid", "offline", "json"] }
solana-program = "~1.16"
blockbuster = "=0.9.0-beta.5"
sha2 = "0.10.0"

[dev-dependencies]
hex = "0.4.3"
