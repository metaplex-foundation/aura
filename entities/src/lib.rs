pub mod enums;
pub mod models;
