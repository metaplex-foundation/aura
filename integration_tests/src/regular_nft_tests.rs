use std::sync::Arc;

use entities::api_req_params::{GetAsset, GetAssetBatch, GetAssetsByGroup, SearchAssets};
use function_name::named;
use itertools::Itertools;
use serial_test::serial;
use tokio::{sync::Mutex, task::JoinSet};

use super::common::*;

#[tokio::test]
#[serial]
#[named]
async fn test_reg_get_asset() {
    let name = trim_test_name(function_name!());
    let setup = TestSetup::new_with_options(
        name.clone(),
        TestSetupOptions { network: None, clear_db: true },
    )
    .await;

    let seeds: Vec<SeedEvent> = seed_nfts(["CMVuYDS9nTeujfTPJb8ik7CRhAqZv4DfjfdamFLkJgxE"]);

    index_seed_events(&setup, seeds.iter().collect_vec()).await;

    let request = r#"        
    {
        "id": "CMVuYDS9nTeujfTPJb8ik7CRhAqZv4DfjfdamFLkJgxE"
    }
    "#;

    let mutexed_tasks = Arc::new(Mutex::new(JoinSet::new()));

    let request: GetAsset = serde_json::from_str(request).unwrap();
    let response = setup.das_api.get_asset(request, mutexed_tasks.clone()).await.unwrap();
    insta::assert_json_snapshot!(name, response);
}

#[tokio::test]
#[serial]
#[named]
async fn test_reg_get_asset_batch() {
    let name = trim_test_name(function_name!());
    let setup = TestSetup::new_with_options(
        name.clone(),
        TestSetupOptions { network: None, clear_db: true },
    )
    .await;

    let seeds: Vec<SeedEvent> = seed_nfts([
        "HTKAVZZrDdyecCxzm3WEkCsG1GUmiqKm73PvngfuYRNK",
        "2NqdYX6kJmMUoChnDXU2UrP9BsoPZivRw3uJG8iDhRRd",
        "5rEeYv8R25b8j6YTHJvYuCKEzq44UCw1Wx1Wx2VPPLz1",
    ]);

    index_seed_events(&setup, seeds.iter().collect_vec()).await;

    for (request, individual_test_name) in [
        (
            r#"        
        {
            "ids": ["HTKAVZZrDdyecCxzm3WEkCsG1GUmiqKm73PvngfuYRNK", "2NqdYX6kJmMUoChnDXU2UrP9BsoPZivRw3uJG8iDhRRd"]
        }
        "#,
            "only-2",
        ),
        (
            r#"        
        {
            "ids": ["2NqdYX6kJmMUoChnDXU2UrP9BsoPZivRw3uJG8iDhRRd", "5rEeYv8R25b8j6YTHJvYuCKEzq44UCw1Wx1Wx2VPPLz1"]
        }
        "#,
            "only-2-different-2",
        ),
        (
            r#"        
        {
            "ids": [
                "2NqdYX6kJmMUoChnDXU2UrP9BsoPZivRw3uJG8iDhRRd",
                "JECLQnbo2CCL8Ygn6vTFn7yeKn8qc7i51bAa9BCAJnWG",
                "5rEeYv8R25b8j6YTHJvYuCKEzq44UCw1Wx1Wx2VPPLz1"
            ]
        }
        "#,
            "2-and-a-missing-1",
        ),
    ] {
        let mutexed_tasks = Arc::new(Mutex::new(JoinSet::new()));

        let request: GetAssetBatch = serde_json::from_str(request).unwrap();
        let response = setup.das_api.get_asset_batch(request, mutexed_tasks.clone()).await.unwrap();
        insta::assert_json_snapshot!(format!("{}-{}", name, individual_test_name), response);
    }
}

#[tokio::test]
#[serial]
#[named]
async fn test_reg_get_asset_by_group() {
    let name = trim_test_name(function_name!());
    let setup = TestSetup::new_with_options(
        name.clone(),
        TestSetupOptions { network: None, clear_db: true },
    )
    .await;

    let seeds: Vec<SeedEvent> = seed_nfts([
        "7jFuJ73mBPDdLMvCYxzrpFTD9FeDudRxdXGDALP5Cp2W",
        "BioVudBTjJnuDW22q62XPhGP87sVwZKcQ46MPSNz4gqi",
        "Fm9S3FL23z3ii3EBBv8ozqLninLvhWDYmcHcHaZy6nie",
    ]);

    index_seed_events(&setup, seeds.iter().collect_vec()).await;

    let request = r#"        
    {
        "groupKey": "collection",
        "groupValue": "8Rt3Ayqth4DAiPnW9MDFi63TiQJHmohfTWLMQFHi4KZH",
        "sortBy": {
            "sortBy": "updated",
            "sortDirection": "asc"
        },
        "page": 1,
        "limit": 1
    }
    "#;

    let mutexed_tasks = Arc::new(Mutex::new(JoinSet::new()));

    let request: GetAssetsByGroup = serde_json::from_str(request).unwrap();
    let response = setup.das_api.get_assets_by_group(request, mutexed_tasks.clone()).await.unwrap();
    insta::assert_json_snapshot!(name, response);
}

#[tokio::test]
#[serial]
#[named]
async fn test_reg_search_assets() {
    let name = trim_test_name(function_name!());
    let setup = TestSetup::new_with_options(
        name.clone(),
        TestSetupOptions { network: None, clear_db: true },
    )
    .await;

    let seeds: Vec<SeedEvent> = seed_nfts([
        "2PfAwPb2hdgsf7xCKyU2kAWUGKnkxYZLfg5SMf4YP1h2",
        "Dt3XDSAdXAJbHqvuycgCTHykKCC7tntMFGMmSvfBbpTL",
    ]);

    index_seed_events(&setup, seeds.iter().collect_vec()).await;

    let request = r#"        
    {
        "ownerAddress": "6Cr66AabRYymhZgYQSfTCo6FVpH18wXrMZswAbcErpyX",
        "page": 1,
        "limit": 2
    }
    "#;

    let mutexed_tasks = Arc::new(Mutex::new(JoinSet::new()));

    let request: SearchAssets = serde_json::from_str(request).unwrap();
    let response = setup.das_api.search_assets(request, mutexed_tasks.clone()).await.unwrap();
    insta::assert_json_snapshot!(name, response);
}

#[tokio::test]
#[serial]
#[named]
async fn test_regular_nft_collection() {
    let name = trim_test_name(function_name!());
    let setup = TestSetup::new_with_options(
        name.clone(),
        TestSetupOptions { network: Some(Network::Mainnet), clear_db: true },
    )
    .await;

    let seeds: Vec<SeedEvent> = seed_nfts(["J1S9H3QjnRtBbbuD4HjPV6RpRhwuk4zKbxsnCHuTgh9w"]);

    index_seed_events(&setup, seeds.iter().collect_vec()).await;

    let request = r#"        
    {
        "id": "J1S9H3QjnRtBbbuD4HjPV6RpRhwuk4zKbxsnCHuTgh9w"
    }
    "#;

    let mutexed_tasks = Arc::new(Mutex::new(JoinSet::new()));

    let request: GetAsset = serde_json::from_str(request).unwrap();
    let response = setup.das_api.get_asset(request, mutexed_tasks.clone()).await.unwrap();
    insta::assert_json_snapshot!(name.clone(), response);
}
