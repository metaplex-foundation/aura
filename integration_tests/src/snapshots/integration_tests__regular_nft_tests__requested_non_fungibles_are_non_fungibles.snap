---
source: integration_tests/src/regular_nft_tests.rs
assertion_line: 271
expression: response
snapshot_kind: text
---
{
  "total": 5,
  "grand_total": 5,
  "limit": 500,
  "items": [
    {
      "interface": "V1_NFT",
      "id": "DvpMQyF8sT6hPBewQf6VrVESw6L1zewPyNit1CSt1tDJ",
      "content": {
        "$schema": "https://schema.metaplex.com/nft1.0.json",
        "json_uri": "https://nftstorage.link/ipfs/bafkreigq3rvy636lvj23tid5ioxd4ovtcivfbhhjst5hk7fhziedygew3a",
        "files": [],
        "metadata": {
          "name": "LIBRE",
          "symbol": "LIBRE"
        },
        "links": {}
      },
      "authorities": [
        {
          "address": "",
          "scopes": [
            "full"
          ]
        }
      ],
      "compression": {
        "eligible": false,
        "compressed": false,
        "data_hash": "",
        "creator_hash": "",
        "asset_hash": "",
        "tree": "",
        "seq": 0,
        "leaf_id": 0
      },
      "grouping": [],
      "royalty": {
        "royalty_model": "creators",
        "target": null,
        "percent": 0.0,
        "basis_points": 0,
        "primary_sale_happened": false,
        "locked": false
      },
      "creators": [],
      "ownership": {
        "frozen": false,
        "delegated": false,
        "delegate": null,
        "ownership_model": "single",
        "owner": "EcxjN4mea6Ah9WSqZhLtSJJCZcxY73Vaz6UVHFZZ5Ttz"
      },
      "supply": {
        "print_max_supply": 0,
        "print_current_supply": 0,
        "edition_nonce": null
      },
      "mutable": true,
      "burnt": false,
      "mint_extensions": {
        "metadata_pointer": {
          "authority": "2jL7yFGXkKE9oi1xHsA45UzV5491PD55AugGJiTbUr9m",
          "metadata_address": "DvpMQyF8sT6hPBewQf6VrVESw6L1zewPyNit1CSt1tDJ"
        },
        "metadata": {
          "update_authority": "2jL7yFGXkKE9oi1xHsA45UzV5491PD55AugGJiTbUr9m",
          "mint": "DvpMQyF8sT6hPBewQf6VrVESw6L1zewPyNit1CSt1tDJ",
          "name": "LIBRE",
          "symbol": "LIBRE",
          "uri": "https://nftstorage.link/ipfs/bafkreigq3rvy636lvj23tid5ioxd4ovtcivfbhhjst5hk7fhziedygew3a",
          "additional_metadata": []
        },
        "group_member_pointer": {
          "authority": "2jL7yFGXkKE9oi1xHsA45UzV5491PD55AugGJiTbUr9m",
          "member_address": "DvpMQyF8sT6hPBewQf6VrVESw6L1zewPyNit1CSt1tDJ"
        }
      },
      "token_info": {
        "supply": 1,
        "decimals": 0,
        "token_program": "TokenzQdBNbLqP5VEhdkAS6EPFLC1PHnBqCXEpPxuEb",
        "freeze_authority": "2jL7yFGXkKE9oi1xHsA45UzV5491PD55AugGJiTbUr9m"
      }
    },
    {
      "interface": "V1_NFT",
      "id": "9qA21TR9QTsQeR5sP6L2PytjgxXcVRSyqUY5vRcUogom",
      "content": {
        "$schema": "https://schema.metaplex.com/nft1.0.json",
        "json_uri": "https://api.akord.com/files/5d0e6fa5-01dc-4893-a2d7-64be4d576449",
        "files": [],
        "metadata": {
          "name": "backpack.eclipse",
          "symbol": "SN"
        },
        "links": {}
      },
      "authorities": [
        {
          "address": "",
          "scopes": [
            "full"
          ]
        }
      ],
      "compression": {
        "eligible": false,
        "compressed": false,
        "data_hash": "",
        "creator_hash": "",
        "asset_hash": "",
        "tree": "",
        "seq": 0,
        "leaf_id": 0
      },
      "grouping": [],
      "royalty": {
        "royalty_model": "creators",
        "target": null,
        "percent": 0.0,
        "basis_points": 0,
        "primary_sale_happened": false,
        "locked": false
      },
      "creators": [],
      "ownership": {
        "frozen": false,
        "delegated": false,
        "delegate": null,
        "ownership_model": "single",
        "owner": "EcxjN4mea6Ah9WSqZhLtSJJCZcxY73Vaz6UVHFZZ5Ttz"
      },
      "supply": {
        "print_max_supply": 0,
        "print_current_supply": 0,
        "edition_nonce": null
      },
      "mutable": true,
      "burnt": false,
      "mint_extensions": {
        "metadata_pointer": {
          "authority": "6qVNPWh4Qhr6pBVZQMNpxwn477rg5LUkjBFuG5KpZvJc",
          "metadata_address": "9qA21TR9QTsQeR5sP6L2PytjgxXcVRSyqUY5vRcUogom"
        },
        "metadata": {
          "update_authority": "6qVNPWh4Qhr6pBVZQMNpxwn477rg5LUkjBFuG5KpZvJc",
          "mint": "9qA21TR9QTsQeR5sP6L2PytjgxXcVRSyqUY5vRcUogom",
          "name": "backpack.eclipse",
          "symbol": "SN",
          "uri": "https://api.akord.com/files/5d0e6fa5-01dc-4893-a2d7-64be4d576449",
          "additional_metadata": []
        },
        "group_member_pointer": {
          "authority": "6qVNPWh4Qhr6pBVZQMNpxwn477rg5LUkjBFuG5KpZvJc",
          "member_address": "9qA21TR9QTsQeR5sP6L2PytjgxXcVRSyqUY5vRcUogom"
        }
      },
      "token_info": {
        "supply": 1,
        "decimals": 0,
        "token_program": "TokenzQdBNbLqP5VEhdkAS6EPFLC1PHnBqCXEpPxuEb"
      }
    },
    {
      "interface": "V1_NFT",
      "id": "8WKGo1z9k3PjTsQw5GDQmvAbKwuRGtb4APkCneH8AVY1",
      "content": {
        "$schema": "https://schema.metaplex.com/nft1.0.json",
        "json_uri": "https://nftstorage.link/ipfs/bafkreigq3rvy636lvj23tid5ioxd4ovtcivfbhhjst5hk7fhziedygew3a",
        "files": [],
        "metadata": {
          "name": "LIBRE",
          "symbol": "LIBRE"
        },
        "links": {}
      },
      "authorities": [
        {
          "address": "",
          "scopes": [
            "full"
          ]
        }
      ],
      "compression": {
        "eligible": false,
        "compressed": false,
        "data_hash": "",
        "creator_hash": "",
        "asset_hash": "",
        "tree": "",
        "seq": 0,
        "leaf_id": 0
      },
      "grouping": [],
      "royalty": {
        "royalty_model": "creators",
        "target": null,
        "percent": 0.0,
        "basis_points": 0,
        "primary_sale_happened": false,
        "locked": false
      },
      "creators": [],
      "ownership": {
        "frozen": false,
        "delegated": false,
        "delegate": null,
        "ownership_model": "single",
        "owner": "EcxjN4mea6Ah9WSqZhLtSJJCZcxY73Vaz6UVHFZZ5Ttz"
      },
      "supply": {
        "print_max_supply": 0,
        "print_current_supply": 0,
        "edition_nonce": null
      },
      "mutable": true,
      "burnt": false,
      "mint_extensions": {
        "metadata_pointer": {
          "authority": "2jL7yFGXkKE9oi1xHsA45UzV5491PD55AugGJiTbUr9m",
          "metadata_address": "8WKGo1z9k3PjTsQw5GDQmvAbKwuRGtb4APkCneH8AVY1"
        },
        "metadata": {
          "update_authority": "2jL7yFGXkKE9oi1xHsA45UzV5491PD55AugGJiTbUr9m",
          "mint": "8WKGo1z9k3PjTsQw5GDQmvAbKwuRGtb4APkCneH8AVY1",
          "name": "LIBRE",
          "symbol": "LIBRE",
          "uri": "https://nftstorage.link/ipfs/bafkreigq3rvy636lvj23tid5ioxd4ovtcivfbhhjst5hk7fhziedygew3a",
          "additional_metadata": []
        },
        "group_member_pointer": {
          "authority": "2jL7yFGXkKE9oi1xHsA45UzV5491PD55AugGJiTbUr9m",
          "member_address": "8WKGo1z9k3PjTsQw5GDQmvAbKwuRGtb4APkCneH8AVY1"
        }
      },
      "token_info": {
        "supply": 1,
        "decimals": 0,
        "token_program": "TokenzQdBNbLqP5VEhdkAS6EPFLC1PHnBqCXEpPxuEb",
        "freeze_authority": "2jL7yFGXkKE9oi1xHsA45UzV5491PD55AugGJiTbUr9m"
      }
    },
    {
      "interface": "V1_NFT",
      "id": "7ZkXycbrAhVzeB9ngnjcCdjk5bxTJYzscSZMhRRBx3QB",
      "content": {
        "$schema": "https://schema.metaplex.com/nft1.0.json",
        "json_uri": "ipfs://QmTmNdx3cNT4wv5yyi7ajvPLZTVia6QzWzg1AAkKtauqqs",
        "files": [],
        "metadata": {
          "name": "Path of Discovery",
          "symbol": "POD"
        },
        "links": {}
      },
      "authorities": [
        {
          "address": "",
          "scopes": [
            "full"
          ]
        }
      ],
      "compression": {
        "eligible": false,
        "compressed": false,
        "data_hash": "",
        "creator_hash": "",
        "asset_hash": "",
        "tree": "",
        "seq": 0,
        "leaf_id": 0
      },
      "grouping": [],
      "royalty": {
        "royalty_model": "creators",
        "target": null,
        "percent": 0.0,
        "basis_points": 0,
        "primary_sale_happened": false,
        "locked": false
      },
      "creators": [],
      "ownership": {
        "frozen": false,
        "delegated": false,
        "delegate": null,
        "ownership_model": "single",
        "owner": "EcxjN4mea6Ah9WSqZhLtSJJCZcxY73Vaz6UVHFZZ5Ttz"
      },
      "supply": {
        "print_max_supply": 0,
        "print_current_supply": 0,
        "edition_nonce": null
      },
      "mutable": true,
      "burnt": false,
      "mint_extensions": {
        "metadata_pointer": {
          "authority": "6wBjjwmWAZzuzAYt9N7EbhPsthE6m2LStNAxZUbViqQb",
          "metadata_address": "7ZkXycbrAhVzeB9ngnjcCdjk5bxTJYzscSZMhRRBx3QB"
        },
        "metadata": {
          "update_authority": "6wBjjwmWAZzuzAYt9N7EbhPsthE6m2LStNAxZUbViqQb",
          "mint": "7ZkXycbrAhVzeB9ngnjcCdjk5bxTJYzscSZMhRRBx3QB",
          "name": "Path of Discovery",
          "symbol": "POD",
          "uri": "ipfs://QmTmNdx3cNT4wv5yyi7ajvPLZTVia6QzWzg1AAkKtauqqs",
          "additional_metadata": [
            [
              "royalty_basis_points",
              "5000"
            ],
            [
              "AsSKqK7CkxFUf3KaoQzzr8ZLPm5fFguUtVE5QwGALQQn",
              "100"
            ]
          ]
        },
        "group_member_pointer": {
          "authority": "6wBjjwmWAZzuzAYt9N7EbhPsthE6m2LStNAxZUbViqQb",
          "member_address": "7ZkXycbrAhVzeB9ngnjcCdjk5bxTJYzscSZMhRRBx3QB"
        }
      },
      "token_info": {
        "supply": 1,
        "decimals": 0,
        "token_program": "TokenzQdBNbLqP5VEhdkAS6EPFLC1PHnBqCXEpPxuEb",
        "freeze_authority": "6wBjjwmWAZzuzAYt9N7EbhPsthE6m2LStNAxZUbViqQb"
      }
    },
    {
      "interface": "V1_NFT",
      "id": "75peBtH5MwfA5t9uhr51AYL7MR5DbPJ5xQ7wizzvowUH",
      "content": {
        "$schema": "https://schema.metaplex.com/nft1.0.json",
        "json_uri": "https://s3.filebase.com/eclipsedomains/eclipse-item.json",
        "files": [],
        "metadata": {
          "name": "backpack.eclipse",
          "symbol": "ED"
        },
        "links": {}
      },
      "authorities": [
        {
          "address": "",
          "scopes": [
            "full"
          ]
        }
      ],
      "compression": {
        "eligible": false,
        "compressed": false,
        "data_hash": "",
        "creator_hash": "",
        "asset_hash": "",
        "tree": "",
        "seq": 0,
        "leaf_id": 0
      },
      "grouping": [],
      "royalty": {
        "royalty_model": "creators",
        "target": null,
        "percent": 0.0,
        "basis_points": 0,
        "primary_sale_happened": false,
        "locked": false
      },
      "creators": [],
      "ownership": {
        "frozen": false,
        "delegated": false,
        "delegate": null,
        "ownership_model": "single",
        "owner": "EcxjN4mea6Ah9WSqZhLtSJJCZcxY73Vaz6UVHFZZ5Ttz"
      },
      "supply": {
        "print_max_supply": 0,
        "print_current_supply": 0,
        "edition_nonce": null
      },
      "mutable": true,
      "burnt": false,
      "mint_extensions": {
        "permanent_delegate": {
          "delegate": "8JGNPBnUFHyjtzF8h9orj7fWNRLJBbvJZvZB2nwjE4Zo"
        },
        "metadata_pointer": {
          "authority": "8JGNPBnUFHyjtzF8h9orj7fWNRLJBbvJZvZB2nwjE4Zo",
          "metadata_address": "75peBtH5MwfA5t9uhr51AYL7MR5DbPJ5xQ7wizzvowUH"
        },
        "metadata": {
          "update_authority": "8JGNPBnUFHyjtzF8h9orj7fWNRLJBbvJZvZB2nwjE4Zo",
          "mint": "75peBtH5MwfA5t9uhr51AYL7MR5DbPJ5xQ7wizzvowUH",
          "name": "backpack.eclipse",
          "symbol": "ED",
          "uri": "https://s3.filebase.com/eclipsedomains/eclipse-item.json",
          "additional_metadata": [
            [
              "royalty_basis_points",
              "500"
            ],
            [
              "FBepPeUcmFupjK7rWonB5gKtjT16AbExnhSVC1bBPAqf",
              "100"
            ],
            [
              "expiration_date",
              "1765018176"
            ],
            [
              "domain",
              "backpack.eclipse"
            ],
            [
              "extension",
              "eclipse"
            ],
            [
              "length",
              "8"
            ],
            [
              "created_at",
              "1733482176"
            ],
            [
              "updated_at",
              "1735340757"
            ]
          ]
        },
        "group_member_pointer": {
          "authority": "8JGNPBnUFHyjtzF8h9orj7fWNRLJBbvJZvZB2nwjE4Zo",
          "member_address": "EdmnsgHGEeeBoGEhtK4th6L3ebmfzfR9Lvw9oi9t4nnL"
        }
      },
      "token_info": {
        "supply": 1,
        "decimals": 0,
        "token_program": "TokenzQdBNbLqP5VEhdkAS6EPFLC1PHnBqCXEpPxuEb",
        "freeze_authority": "8JGNPBnUFHyjtzF8h9orj7fWNRLJBbvJZvZB2nwjE4Zo"
      }
    }
  ],
  "cursor": "75peBtH5MwfA5t9uhr51AYL7MR5DbPJ5xQ7wizzvowUH"
}
